### ISSUE TITLE
First thoughts & make test-all

### ISSUE THREAD

**Issue by MSP-Greg** [oneclick/rubyinstaller2#3]
First of all, thank you and a wonderful start.  I could even (maybe?) help with maintenance, even though I'm not a c guy...

'First pass thru' items:

1. right below [line #56 in PKGBUILD](https://github.com/larskanis/rubyinstaller2/blob/master/package/ruby-2.4.0/PKGBUILD#L56), I added --
```
make -l "TESTOPTS=-v -q" test-all > "D:\make_test-all.txt"
```
result was --
```
16496 tests, 2226300 assertions, 14 failures, 182 errors, 151 skips
```
Many were in `gdbm`, probably due to item 2.  I did comment out the [lines in test_enum.rb](https://github.com/ruby/ruby/blob/trunk/test/ruby/test_enum.rb#L559-L583) mentioned over in RubyInstaller, as without them commented out, I had a segfault.  That segfault also happened with my msys2 / gcc 6.2.0 build based on RubyInstaller.

2. Std-Lib - dbm not installed

Here's what I have for gdbm packages --
```
$ pacman -Ss gdbm
mingw32/mingw-w64-i686-gdbm 1.12-1 [installed]
    GNU database library (mingw-w64)
mingw64/mingw-w64-x86_64-gdbm 1.12-1 [installed]
    GNU database library (mingw-w64)
msys/gdbm 1.11-3 (Database) [installed]
    GNU database library
msys/libgdbm 1.11-3 (libraries) [installed]
    GNU database library
msys/libgdbm-devel 1.11-3 (development) **not installed**
    libgdbm headers and libraries
```

3. In my `package\ruby-2.4.0 folder`, I have `ruby-2.4.0.tar.bz2`, it also exists in the `package\ruby-2.4.0\src` folder.  Maybe that's needed...

4. It would be helpful if one could skip the RDoc/ri step.  Then again, I'm somewhat opinionated about docs, and I want to review code on a tablet.  See [msp-greg.github.io](https://msp-greg.github.io/).

5. Obviously, you just got started, but local and other branches/tags would be helpful.

I would hope that if everyone can help get the `make test-all` issues down to a small and repeatable set, maybe the Ruby-Core team will be willing to test and work (more) on Windows issues, along with some of the gem authors.

Thanks again...

**Comment by larskanis** [comment]
Thanks, @MSP-Greg for your valuable feedback!

My test run shows slightly lower numbers:
```
16492 tests, 2228971 assertions, 12 failures, 95 errors, 153 skips
```

89 of them are related to gdbm and seem to be caused by the same root cause. 18 remaining issues aren't too bad. However my current focus is to bring RubyInstaller2 at the same level of features and stability as RubyInstaller first. I'll work on the tests before the first (non-pre) release.

Each (pre-)release will get a tag in git. The master branch is currently enough for me. But you're free to make your own branches, of course.

**Comment by MSP-Greg** [comment]
Update - I finally created some sym links between my drives, test-all is now 12 failures, 1 error , using Ruby trunk (2107-01-16), 32bit (OS is 64), openssl 1.1.0c, and zlib 1.2.10.  I think I can use RI2 without modification now (due to my multiple drives, and msys64 & GitHub being on d).

Three or four failures may be bugs (I don't know the nomenclature of the 'Ruby Tracking system') :
```
Bug::Win32::TestDln#test_check_imported [Bug #6303]
TestDir_M17N#test_entries_compose [ruby-core:48745] [Bug #7267]
TestFile#test_utime [ruby-core:44776]
TestRubyOptions#test_command_line_progname_nonascii [ruby-dev:48752] [Bug #10555]
```
Been seeing if I could get gdbm 1.12 to build, but haven't gotten it too.

Question - did you need to comment out the two methods/tests in [test_enum.rb](https://github.com/ruby/ruby/blob/trunk/test/ruby/test_enum.rb#L559-L583)?  I need to recheck if they still cause a segfault...

Finally, thanks for all of your work.

**Comment by MSP-Greg** [comment]
@larskanis thanks for all the work getting Appveyor running.  I'm still bouncing back and forth between RI & RI2.  Part of the reason is that I'm using RI with my local repo, and building 2.5.0, which might as well be the stable release, as nothing's been backported to 2.4.  Another reason is that I wanted to try to update the dependencies, as some are rather old.

So I dl'd Appveyor Build 1.0.61 (32 bit), and got the following:
```
GDBM version 1.12. 16/05/2016 (built May 27 2016 12:37:41)
OpenSSL 1.0.2g  1 Mar 2016
Psych (libyaml) 0.1.6
Zlib 1.2.8
```
I'm sure you've noticed, but mingw gdbm 1.12 is built without `--enable-libgdbm-compat`, which I believe breaks it for Ruby.  I tried to build it, had issues, and decided to start back at 1.10, which is at least a few years old, as opposed to 1.8.3 (2002!) in RI.

My current build is running:
```
GDBM version 1.10. 13/11/2011 (built Jan 18 2017 18:39:10)
OpenSSL 1.1.0c  10 Nov 2016
Psych (libyaml) 0.1.7
Zlib 1.2.11
```

I believe I'm still at 12/1 in test-all.  Thanks again.

**Comment by larskanis** [comment]
@MSP-Greg Currently I don't work on Ruby-2.5, because I want to focus to get a stable RubyInstaller2 based on Ruby-2.4 out, soon. However you're free to fork and add a `compile/ruby-2.5-dev` directory, or so.

I don't want to use more recent versions than the one provided by the MSYS2 project. If we want to use newer packages, the way should be to send a pull request to https://github.com/Alexpux/MINGW-packages . You seem to have success to build newer versions of openssl and libyaml with MSYS2. Would you mind to open pull requests to MSYS2 to update these? Then we'll use newer versions in RI2.

Regarding GDBM we could either use our own PKGBUILD for version 1.10 or fix ruby in order to work with the latest version. I didn't put any efforts into it, yet.

Regarding `make test-all` I didn't get any segfaults, but the appveyor run breaks while [testing openssl](https://ci.appveyor.com/project/larskanis/rubyinstaller2/build/1.0.65/job/5r5j0n2j59ch8l8x#L1732) . On my local Windows-10 installation it runs through as described [above](https://github.com/larskanis/rubyinstaller2/issues/3#issuecomment-271219495) .

**Comment by MSP-Greg** [comment]
@larskanis 
> You seem to have success to build newer versions of openssl and libyaml with MSYS2.

Missing word is some.  OpenSSL is generating four of my twelve failures.

I don't think I've got the patches being applied to Ruby (using my RI version), I need to fix that.  Are your patches (in RI2) the same as what's in [MINGW-packages/mingw-w64-ruby](https://github.com/Alexpux/MINGW-packages/tree/master/mingw-w64-ruby)?  Also, I'm not sure about the [libffi-3.2.1-mswin.patch](https://github.com/ruby/ruby/blob/trunk/ext/fiddle/win32/libffi-3.2.1-mswin.patch)  in Ruby...

> Would you mind opening pull requests to MSYS2 to update these?

I'm trying to get all the ruby 'dependencies' updated and working, then I'll submit to MINGW-packages.  Remember, I'm new to this and I don't code in c or c++.  All I can test with are the 'built-in' tests and Ruby.

Re 2.5/trunk, I'm just working with it (as stated above) in the hope that Ruby Core might pay attention if test-all was repeatable and showed a minimal number of issues.  That does bring up an issue with the naming in the files from https://cache.ruby-lang.org/pub/ruby, where some are named 'numerically' and some not.  I think that causes some issues with RI & RI2.

> I don't want to use more recent versions than the one provided by the MSYS2 project.  -and- Regarding GDBM we could either use our own PKGBUILD for version 1.10 or fix ruby

The question I have is what versions are in common use on non-Windows systems?  I have no idea.

I suspect that is what will determine whether Ruby core will support updating 'dependency versions'.

> I didn't put any efforts into it, yet.

Key distinction between your work and mine.  I can find the issues and maybe find work-arounds, but you stand a fighting chance of fixing them at the source.  Thanks again.

**Comment by MSP-Greg** [comment]
@larskanis 

> Regarding GDBM we could either use our own PKGBUILD for version 1.10

I took a brief look at it, gdbm 1.8.3 was released in 2002, 1.10 in 2011.  gdbm 1.11+ involves quite a few changes, more and different shell commands, etc.  IOW, I think my limit is getting 1.10 working.

I'm working on openssl 1.1.0c.  Seems to have quite a bit more ipv6 build in, and other changes.  I'm reviewing the patches, etc.  The version I'm using I created with a modified RI [knap-build](https://github.com/oneclick/knap-build) recipe; I'll try to get a MINGW-packages version working...

**Comment by MSP-Greg** [comment]
@larskanis Thanks again for all your work.

I'm still building with RI (since I prefer to use my git Ruby files), but I've since done a bit with msys2 packaging, and all my RI dependencies are now built with `makepkg-mingw`.  Full `make test-all` results at gist [Ruby test-all 25Feb](https://gist.github.com/MSP-Greg/d07c6e42caa545bcff0407cd50197e75).  See comment at the bottom for summary.

Notes:
* readline 'native' builds, and passes most tests, irb seems to work fine.
* gdbm - built 1.12 with `--enable-libgdbm-compat`, but tests don't seem to be windows 'friendly'.  Didn't work with Ruby.  Using 1.10 with no Ruby test failures.  In package, 17 tests were successful, 5 tests were skipped (Cloexec).
* openssl - using 1.1.0f, same Ruby failures as previous versions, including 1.0.2*.  All OpenSSL tests passed.  Also used config option `enable-capieng`, which may be the better thing to do with Windows.  According to [OpenSSL Compilation and Installation - Configure Options](https://wiki.openssl.org/index.php/Compilation_and_Installation#Configure_Options), it enables the Microsoft CAPI engine on Windows platforms.
* libyaml - updated to 0.1.7

So --
1. What results are you seeing in test-all?
2. Are you interested in my packaging files?  I could start a repo for them.  I think there is the possibility that Ruby may always need some packages build differently than the mys2 packages.
3. I think I'll take a stab at adding a 'local repo' option to RI2.  Do you have a preference for whether it's in Ruby or PKGBUILD?

**Comment by larskanis** [comment]
>1.    What results are you seeing in test-all?

Didn't test for a while.

>2.    Are you interested in my packaging files? I could start a repo for them. I think there is the possibility that Ruby may always need some packages build differently than the mys2 packages.

Yes, of course I'm interested! I prefer to use the msys2 upstream packages whenever possible. But as a interim solution it's OK to add it locally. In the end I would always try to make a pull request upstream for the changes we need or find useful.

I'll also open a pull request for the changes I've done to ruby PKGBUILD, that are interesting upstream, when I find some time.

>3.    I think I'll take a stab at adding a 'local repo' option to RI2. Do you have a preference for whether it's in Ruby or PKGBUILD?

There's not much to do. Just add the PKGBUILD directory in `compile` and add another `makepkg-mingw` run to `task.rake`. And then add another package file to be installed in `unpack`. I plan to make this more configurable in the future (which MSYS package is required by which `RubyPackage`), but for now it's working.
I prefer PKGBUILD.

**Comment by MSP-Greg** [comment]
@larskanis 

> And then add another package file to be installed in `unpack`.

I'm not being clear about what I mean by 'local'...

I'm thinking of a command line like:

`rake local --path /github/ruby --branch trunk`
or
`rake local --path /github/ruby --branch ruby_2_4`

So there's no unpacking to do, maybe copying files so patches are applied in the `src` folder, rather than the user's git repo files.

So, the reason I asked about code in Ruby vs PKGBUILD, is if we allow RI2 to switch branches (rather than the user doing so with git), it might be easier to do so in Ruby, then open the version.h file to check for data needed for build variables.

There's also the issue of loading bundled gems into the src folder, which don't exist in the [ruby/ruby](https://github.com/ruby/ruby) repo, but do exist in the ruby-lang.org files, like https://cache.ruby-lang.org/pub/ruby/2.4/ruby-2.4.0.tar.bz2.

**Comment by larskanis** [comment]
Sorry I thought you want to add customized builds of openssl, readline, etc. to RI2. This is still an option, if upstream packages don't work.

There are a lot of `mingw-w64-*-git` packages in https://github.com/Alexpux/MINGW-packages . They use the code directly from a git repo. This should be possible with Ruby as well. PKGBUILD has all required functions builtin or can use other msys2 packages.

To be honest, I don't have interest in using Ruby from git. I don't want to do Ruby MRI development - I want to use stable code only. But if it's valuable for you, feel free to add it as PKGBUILD. However I don't want to add any Ruby code for Ruby-from-git handling, for now.

**Comment by MSP-Greg** [comment]
@larskanis 

> Sorry I thought you want to add customized builds of openssl, readline, etc. to RI2. This is still an option, if upstream packages don't work.

If they'll accept packages specific to Ruby builds, then there's no issue.  I'll try to get my package code up today or tomorrow so you can test it when you have a minute.

Re the git or git/local, I understand your viewpoint.  I guess I was thinking/hoping that if --

RI2 allowed people to easily build against the git trunk or stable branches =>
more people testing and finding bugs =>
more people reporting to Ruby Core =>
more Ruby Core support of windows =>
test-all is successful  on windows =>
maybe they might even build/test against mingw/msys2, rather than MSVC.

Then again, maybe I'm crazy...

In the mean time, I'll get a PKGBUILD working with local, based on your code, and see if the test-all is better or the same as the test-all I'm getting with my RI version.  BTW, my RI version only has the dependencies built from my packages (essentially the `depends` array in PKGBUILD), the other build files I've got sym-linked to my msys2 folders.  Same for gcc, etc...

**Comment by MSP-Greg** [comment]
@larskanis 

Well, I'm back to my third question.  Thinking about how we all got here, which may have started with the call for contributors to RI.  I'm also reminded of @jonforums's post in #10 --

> When you regularly build ruby from source (e.g. - build from new backports to ruby_2_3 rather than waiting for official releases), it is much more streamlined to be able to build from a local repo,

I understand your thoughts about using stable code only, but in thinking about the future, RI2 may need contributors to keep it current.  So, giving consideration to the feature set that they want may be worthwhile.

If the only people using RI2 are not really using it, but simply downloading a packaged Ruby installer, the base of capable people willing to help is much smaller.  Also, there's the line of thought I mentioned, in terms of better feedback to the Ruby Core team if 'capable people' can build against stable branches or trunk.

Granted, I'm sure being based on msys2/mingw will be very helpful in terms of adoption and contributions, but, simply put, the more people really use RI2, the more likely they are to be willing to contribute...

**Comment by MSP-Greg** [comment]
@larskanis  Thanks for all the work in the gem branch.

I'm still building with trunk, etc, etc.  First, stop seg faults, then minimize skips, then failures/errors.  For some of the seg faults, I needed different patch files for 32 vs 64 bit.

A lot of skips were due to `git` not being available for rubygems tests.  Wrestled a bit with that, but got it pointed to my 'Git for Windows' install, which installs to `Program Files`.  Now, there's all sorts of issues like:

```
fatal: not in a git directory
fatal: Not a git repository
fatal: Could not read from remote repository
fatal: 'E:/temp/test_rubygems_4884/git/a' does not appear to be a git repository
```

Have you come across this?  If not, I may ask in [rubygems / rubygems](https://github.com/rubygems/rubygems)...

**Comment by MSP-Greg** [comment]
@larskanis FYI, filed PR's in rubygems & ruby.  Accepted in rubygems, not yet in ruby.

ENV['GIT'] needs to be set to the full path of git.exe.  Requires path with no spaces.   I have no idea where it would be on appveyor.

**Comment by larskanis** [comment]
Thank you @MSP-Greg for continually fighting against ruby bugs on Windows! I didn't notice this particular issue. That's probably because I still ignore all errors that come from other but my own test suite...

**Comment by MSP-Greg** [comment]
@larkanis FYI, trunk is now at
```
16625 tests, 2234416 assertions, 8 failures, 1 errors, 113 skips
```
But --
1.  I've got 5 patches for segfaults, although some may only be needed for 32 bit (have to get back to checking that).
2. readline is acting up, as irb is choppy.
3. Now using a stable branch build of OpenSSL 1.1.0f, 15 Mar 2017

**Comment by MSP-Greg** [comment]
@larskanis Wow.  Thanks for all the work. Looks great, but when I tried it running rake -T, had an error.  Need to take another look, as I was not in a position to give it my full attention.

Dumb question from a GCC newbie.  [GCC x86 Windows Options](https://gcc.gnu.org/onlinedocs/gcc-6.3.0/gcc/x86-Windows-Options.html#x86-Windows-Options) show a few flags (`-municode`, `-mwin32`) that I've wondered about.  Might these have any (positive) affect on building?  The reason I ask is that some of the failures/errors I have seem to be encoding related, and I did see your UTF-8 patch....

**Comment by larskanis** [comment]
> when I tried it running rake -T, had an error. 

Try `bundle install` and `bundle exec rake -T`. 

No, I don't think that the GCC Windows Options improve anything. They are already set by the ruby build as necessary.

My UTF-8 patch changes the `Encoding.default_external` as [described in the CHANGELOG](https://github.com/larskanis/rubyinstaller2/blame/master/CHANGELOG.md#L8). This is something that I consider as a major interoperability issue and as a very useless setting, because nobody seriously uses the ancient cpXYZ encodings for file content. If a native encoding shall be used, it should be UTF-16 (which is technological nonsense by itself, IMHO) on Windows. However UTF-16 would make interoperability and compatibility even more difficult. So the only reliable choice for `default_external` is UTF-8, IMO.

Nevertheless I need to discuss this with the ruby core team. I also noticed the additional failures in `test-all`. These tests needs to be adjusted accordingly.

**Comment by MSP-Greg** [comment]
@larskanis Thanks, that worked.

I'm working on some code to log `test-all` results to my GitHub.io site, possibly along with with a listing of all the patch files needed.  Maybe I'm dreaming, but ruby core team could then see the status of MinGW builds.  Once that's done, I'll get it working on the stable branches.

I'll also see how RI2 builds do with my packages and SEGV patches on test-all.  If that works, I'll see if they'll be accepted at [Alexpux/MINGW-packages](https://github.com/Alexpux/MINGW-packages) and [SourceForge.net](https://sourceforge.net/projects/msys2/files/REPOS/MINGW/).

Re UTF-8, almost all of my ruby scripts get run with `-Eutf-8:utf-8`, so I agree.

I'm also trying to integrate your private assembly code into my code.  Thanks for that.

1)  I noticed some of the scripts in `/bin` have `/mingw64/bin/` and `/usr/bin/env` in the `#!` lines.  Sometime, somewhere that caused an issue for me.

2)  I haven't quite wrapped my head around everything you've added/changed, but with all the rake files, I think I need to add code to my YARD plug-in to doc them...

**Comment by larskanis** [comment]
The ruby core team already uses [appveyor for CI](https://ci.appveyor.com/project/ruby/ruby/branch/trunk). However they run the basic tests only, not test-all.

Nevertheless appveyor is easy to setup. Just fork the ruby git repository and enable it at your appveyor account. Then tune your git content until you're satisfied with the appveyor result. The status is already embedded into the README and communicated per email.

When the first release of RubyInstaller2 is out, I'm open to add ruby from git to the repository. However I don't really care about the ruby core test results in RubyInstaller2. The important tests in this repository are those in the [test directory](https://github.com/larskanis/rubyinstaller2/tree/master/test). The ruby core tests belong into the ruby repository itself (or the ruby-spec as an independent party) and needs to be tested there. So each party should test the code that it controls.

**Comment by MSP-Greg** [comment]
@larskanis Thank you again for all your work.  After adding your manifest code to my 'build trunk/stable' system, trunk is testing as follows:

```
17090 tests, 4976901 assertions, 8 failures, 1 errors, 111 skips
ruby -v: ruby 2.5.0dev (2017-04-08 trunk 58275) [x64-mingw32]
```

(`assertions` essentially doubled recently)

I first thought I'd just set it up for ruby_2_4 and trunk, but today I decided to try ruby_2_3.  Results:

```
15666 tests, 2221205 assertions, 5 failures, 1 errors, 113 skips
ruby -v: ruby 2.3.4p301 (2017-03-30 revision 58213) [x64-mingw32]
```

I'm using  a handful of patches for test files causing segv issues, etc.  Also, ruby_2_3 wouldn't build with OpenSSL 1.1.0f; I had to push back to 1.0.2k.  I have to revise my code, as I need to add version specific dll info.  I forgot about the file name changes to OpenSSL...

Importantly, re failures/errors, one is with `readline`, all others appear to be 'Windows' related.  None with gdbm, dbm, OpenSSL, or other 'extensions'.  Still using gdbm 1.10, I can't get 1.11 or 1.12 (or 1.13) to build correctly.

Sometime in the somewhat near future I'll put the code up on GitHub.

Thanks again.
